### ISSUE TITLE
ccode excessively slow when using user_functions

### ISSUE THREAD

**Issue by nathanielvirgo** [sympy/sympy#15377]
The following code takes 28 seconds to run:

```
import sympy as sy
x = sy.Symbol('x')
foo = sy.Function('foo')
sy.ccode(foo(foo(foo(foo(foo(foo(foo(foo(foo(foo(foo(foo(foo(x))))))))))))), user_functions={'foo':'foo'})
```

This seems excessively slow. It seems to scale exponentially with the number of times the user function is called, taking around three times longer with each one.

This is a bit of a pain, as I am trying to use this system for code generation. Is there a workaround?

Some observations: 
 * This happens only when using a user-defined function in this way. With a builtin such as `sin`, it's practically instantaneous. 
 * It's also fast if I don't specify ` user_functions={'foo':'foo'}`, but unfortunately in that case it will not correctly process the arguments to the user function, outputting `foo(sy.Min(x,y))` as `'foo(Min(x,y))'` instead of `'foo(fmin(x,y))'`, which is a problem for my application.

**Comment by nathanielvirgo** [comment]
I found a workaround for this issue, in case this affects someone else. Instead of using `Function('foo')`, make foo a subclass of sympy.Function that defines the following method:

```
	def _ccode(self, printer):
		return f"{type(self).__name__}({', '.join(printer._print(arg) for arg in self.args)})"
```

Then it works as quickly as a built-in function. It would still be good to have this fixed, though.

**Comment by moorepants** [comment]
FYI this runs fast in SymPy 1.1.1 and slow on the master branch.

**Comment by moorepants** [comment]
Some information. pyinstrument profile from SymPy 1.1.1:
![selection_006](https://user-images.githubusercontent.com/276007/46811127-fe146180-cd26-11e8-9e8a-447508328a31.png)
and from the master branch:
![selection_007](https://user-images.githubusercontent.com/276007/46811201-21d7a780-cd27-11e8-8cb0-a8cd99b09d0d.png)

**Comment by moorepants** [comment]
Git bisect tells us this is the commit that slowed things down:

```
b4940ae4d936da612971961b55bb5f6d5b451b5f is the first bad commit
commit b4940ae4d936da612971961b55bb5f6d5b451b5f
Author: Björn Dahlgren <bjodah@gmail.com>
Date:   Thu Dec 7 17:15:18 2017 +0100

    Add support for modf in ccode
```

Which is from this PR https://github.com/sympy/sympy/pull/13692, FYI @bjodah.

**Comment by bjodah** [comment]
Ah, I did not foresee that. I won't have time to address this anytime soon, but we should definitely speed this up by changing the control flow.

**Comment by asmeurer** [comment]
I don't fully understand what is happening there. Can this be fixed by reverting that PR and implementing `_print_Mod` instead? Is the problem that the known_functions cannot handle parenthesize natively, because they cannot access `self`?

**Comment by moorepants** [comment]
I would vote for creating specific methods for functions if they require any logic. The `known_functions` dict should probably just be for simple text replacements.

**Comment by moorepants** [comment]
I proposed a fix in #15379.
